bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +840,208 @@ protected void setBusy(boolean busy) {
+	 * @since 3.31
+	 */
+	protected void handleExpandableNodeClicked(Widget cell) {
+		// default implementation does nothing. Actual viewers can decide how to
+		// populate remaining elements.
+	}
+
+	/**
+	 * If the items limit is not set, or the number of visible children is below the
+	 * items limit, this method does nothing and returns {@code null}.
+	 * <p>
+	 * In other case, this method tries to fetch (probably updated) model elements
+	 * up to the number of given visible elements, expanding possible existing
+	 * {@link IExpandableNode} elements if needed.
+	 * <p>
+	 * Implementation note: we should not dispose already visible items and we
+	 * should display limited elements along with already visible items.
+	 *
+	 * @param parent          model item to be refreshed.
+	 * @param visibleChildren currently visible children. This includes any elements
+	 *                        already expanded by user.
+	 * @return list of children to be displayed/refreshed or {@code null} if given
+	 *         items length does not exceed items limit
+	 */
+	Object[] getChildrenWithLimitApplied(final Object parent, Item[] visibleChildren) {
+		final int limit = getItemsLimit();
+		final int visibleItemsLength = visibleChildren.length;
+		if (visibleItemsLength <= limit || limit <= 0) {
+			return null;
+		}
+		// if we are here, at least one iteration of expandable items is shown
+
+		Object[] currentSorted = getSortedChildren(parent);
+		// if current children count is below limit just return them
+		if (currentSorted.length < limit) {
+			return currentSorted;
+		}
+
+		// numberOfTimesRead is equal to number of times it was expanded by user.
+		int numberOfTimesRead = visibleItemsLength / limit;
+		if ((visibleItemsLength % limit) > 0
+				&& !(visibleChildren[visibleItemsLength - 1].getData() instanceof IExpandableNode)) {
+			numberOfTimesRead++;
+		}
+		List<Object> newChildren = new ArrayList<>();
+		Object lastElement = parent;
+		while (numberOfTimesRead > 0) {
+			Object[] sortedChildren;
+			if (lastElement == parent) {
+				sortedChildren = currentSorted;
+			} else {
+				sortedChildren = getSortedChildren(lastElement);
+			}
+			lastElement = sortedChildren[sortedChildren.length - 1];
+			Collections.addAll(newChildren, sortedChildren);
+
+			if (!(lastElement instanceof IExpandableNode)) {
+				// no more to read: less items than visibleChildren but more than limit.
+				break;
+			}
+			numberOfTimesRead--;
+			if (numberOfTimesRead > 0) {
+				// remove expandable node from the end of the list
+				// we will expand it at least once more on next loop round
+				newChildren.remove(newChildren.size() - 1);
+			}
+		}
+		return newChildren.toArray(new Object[newChildren.size()]);
+	}
+
+	/**
+	 * Returns the current viewer limit on direct children at one level. Limit that
+	 * is less than or equal to zero has no effect on the viewer.
+	 *
+	 * @return current limit
+	 * @since 3.31
+	 */
+	public int getItemsLimit() {
```

Comment:
Would it be possible to keep is package protected (thus not API)?

Reply:
now `public int getItemsLimit()` is only package protected.